internals/overlord/certstate/manager.go

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// certstate manages key-pair and certificate creation and selection.
+package certstate
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"io/fs"
+	"math/big"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+type CertManager struct {
+	tlsDir           string
+	x509KeyPairsLock sync.Mutex
+	x509KeyPairs     []x509KeyPair
+	changeListeners  []CertChangedFunc
+}
+
+// x509KeyPair represents the X509 public certificate and the private
+// key that signed it.
+type x509KeyPair struct {
+	Order       int
+	Private     any
+	Certificate *x509.Certificate
+}
+
+func NewManager(tlsDir string) (*CertManager, error) {
+	manager := &CertManager{tlsDir: tlsDir}
+	if err := manager.loadX509KeyPairs(); err != nil {
+		return nil, err
+	}
+	return manager, nil
+}
+
+// X509Keypair returns a valid tls.Certificate instance that
+// can be used for starting a TLS based network listener. A call
+// to this function may generate a new X509 keypair if none
+// of the available pairs are valid.
+func (c *CertManager) X509Keypair() (*tls.Certificate, error) {
```

Comment:
Add the thread that creates the candidate certificates, before the current one expires.